File: leetcode/0155-min-stack/solution.cpp
class MinStack {
public:
    int minElement;
    stack<long long> s;
    
    /** initialize your data structure here. */
    MinStack() {
        minElement = INT_MAX;
    }
    
    void push(int val) {
        s.push(2L * val - minElement);
        
        if (val < minElement) {
            minElement = val;
        }
    }
    
    void pop() {
        long long top = s.top();
        s.pop();
        
        if (top < minElement) {
            minElement = (int)(2L * minElement - top);
        }
    }
    
    int top() {
        long long top = s.top();
        
        if (top < minElement) {
            return minElement;
        } else {
            return (int) ((top + minElement) / 2);
        }
    }
    
    int getMin() {
        return minElement;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: leetcode/0155-min-stack/solution.py
class MinStack:
    
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = list()
        self.minElement = float(inf)

    def push(self, val: int) -> None:
        self.stack.append(2 * val - self.minElement)
        
        if val < self.minElement:
            self.minElement = val

    def pop(self) -> None:
        top = self.stack.pop()
        
        if top < self.minElement:
            self.minElement = 2 * self.minElement - top

    def top(self) -> int:
        top = self.stack[-1]
        
        if top < self.minElement:
            return self.minElement
        else:
            return (top + self.minElement) // 2

    def getMin(self) -> int:
        return self.minElement


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()